device/devicegate/filter.go

```diff
@@ -0,0 +1,263 @@
+package devicegate
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/xmidt-org/webpa-common/device"
+	"github.com/xmidt-org/webpa-common/logging"
+	"github.com/xmidt-org/webpa-common/xhttp"
+)
+
+const (
+	metadataMapLocation = "metadata_map"
+	claimsLocation      = "claims"
+)
+
+var (
+	emptyVal = struct{}{}
+)
+
+type FilterGateOption func(*FilterGate)
+
+type FilterGate struct {
+	filters        map[string]map[interface{}]struct{}
+	allowedFilters map[string]struct{}
+	lock           sync.RWMutex
+}
+
+type filterRequest struct {
+	Key    string
+	Values []interface{}
+}
+
+func NewFilterGate(options ...FilterGateOption) *FilterGate {
+	fg := &FilterGate{
+		filters:        make(map[string]map[interface{}]struct{}),
+		allowedFilters: make(map[string]struct{}),
+	}
+
+	for _, o := range options {
+		o(fg)
+	}
+
+	return fg
+
+}
+
+func WithAllowedFilters(allowedFilters map[string]struct{}) FilterGateOption {
+	return func(fg *FilterGate) {
+		if allowedFilters != nil {
+			fg.allowedFilters = allowedFilters
+		} else {
+			fg.allowedFilters = make(map[string]struct{})
+		}
+	}
+}
+
+func (f *FilterGate) PrettifyFilters() map[string][]interface{} {
+	copy := make(map[string][]interface{})
+
+	f.lock.RLock()
+	for k, v := range f.filters {
+		var filters []interface{}
+		for filter := range v {
+			filters = append(filters, filter)
+		}
+
+		copy[k] = filters
+	}
+	f.lock.RUnlock()
+
+	return copy
+}
+
+func (f *FilterGate) FiltersCopy() map[string]map[interface{}]struct{} {
+	filtersCopy := make(map[string]map[interface{}]struct{})
+
+	f.lock.RLock()
+	for key, val := range f.filters {
+		valCopy := make(map[interface{}]struct{})
+
+		for k, v := range val {
+			valCopy[k] = v
+		}
+
+		filtersCopy[key] = valCopy
+	}
+	f.lock.RUnlock()
+
+	return filtersCopy
+}
+
+func (f *FilterGate) Filters() map[string]map[interface{}]struct{} {
+	return f.filters
+}
+
+//allows for adding and removing filters
+//Will completely overwrite current filters in place
+func (f *FilterGate) EditFilters(key string, values []interface{}, add bool) bool {
```

Comment:
See my earlier comment about not exposing the filters map.  You can supply an API that's easier to use with methods like:

```
Set(key string, value ...interface{}) []interface{} // return the old values, you may also want to return a bool
Delete(key string)
```